File: surface/widgets/winui/components/widgets/primitives/VisuallyHidden.cs
// ============================================================
// Clef Surface WinUI Widget — VisuallyHidden
//
// Renders content that is invisible to sighted users but
// accessible to screen readers and Narrator. In WinUI 3 this
// is achieved by collapsing the element visually while keeping
// the automation name set for accessibility services.
//
// Adapts the visually-hidden.widget spec: anatomy (root),
// states (static), and connect attributes (data-part, style)
// to WinUI 3 rendering.
// ============================================================

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;

namespace Clef.Surface.WinUI.Widgets.Primitives;

public sealed class ClefVisuallyHidden : UserControl
{
    public static readonly DependencyProperty TextProperty =
        DependencyProperty.Register(nameof(Text), typeof(string), typeof(ClefVisuallyHidden),
            new PropertyMetadata(null, OnPropertyChanged));

    public string Text { get => (string)GetValue(TextProperty); set => SetValue(TextProperty, value); }

    private readonly Border _container;

    public ClefVisuallyHidden()
    {
        _container = new Border
        {
            Width = 0,
            Height = 0,
            Opacity = 0
        };
        Content = _container;
        UpdateVisual();
    }

    public void SetHiddenContent(UIElement element)
    {
        _container.Child = element;
    }

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        => ((ClefVisuallyHidden)d).UpdateVisual();

    private void UpdateVisual()
    {
        if (Text != null)
            AutomationProperties.SetName(_container, Text);
    }
}

File: surface/widgets/nativescript/components/widgets/primitives/VisuallyHidden.ts
// ============================================================
// Clef Surface NativeScript Widget — VisuallyHidden
//
// Hides content visually while keeping it accessible to screen
// readers. Sets zero dimensions and clips overflow.
// ============================================================

import { StackLayout } from '@nativescript/core';
import type { View } from '@nativescript/core';

// --------------- Props ---------------

export interface VisuallyHiddenProps {
  children?: View[];
}

// --------------- Component ---------------

export function createVisuallyHidden(props: VisuallyHiddenProps): StackLayout {
  const { children = [] } = props;

  const container = new StackLayout();
  container.className = 'clef-widget-visually-hidden';
  container.width = 1;
  container.height = 1;
  container.opacity = 0;
  container.clipToBounds = true;

  for (const child of children) {
    container.addChild(child);
  }

  return container;
}

export default createVisuallyHidden;
